### ISSUE TITLE
How do you encode a space in a Phrase

### ISSUE THREAD

**Issue by kcking** [1aim/mail#14]
Hi there, I'm trying to encode a source `Mailbox` to send to the AWS SES API, but the display name (represented as a `Phrase`) isn't encoding any spaces between words. I'm instantiating an `EncodingBuffer`, then calling `to_string()`, but I noticed in `mail_headers/src/header_components/phrase.rs:44` there's a FIXME: `currently collapses WS`. Sure enough, AWS just sets the From name to the concatenation of each word without spaces in between.

Is there any workaround to this issue?

**Comment by kcking** [comment]
I found a workaround by replacing spaces with underscores in the input, not sure if this is intended behavior though

**Comment by rustonaut** [comment]
TL;DR: use `Phrase::new` or `Phrase::try_from` (which accepts `&str`,`String`)

---

From your error I guess that you create a `Phrase` instance manually from `Word` instances.
(If not there is a bug somewhere and the following doesn't really apply).

The thing is word instances are based on the mail grammars word part. Which mean they include white spaces (and might actually not be words but just parts of word or whole sentences as quoted strings).

So you would need to use e.g. `Word::pad_left`/`Word::pad_right` if you want to do it that way. But I strongly recommend you to just use `Phrase::new` or `Phrase::try_from`. (E.g. `Phrase::new("Some thing  fun")?`).

Note that `Phrase::new` (and `try_from`) can fail if you have a empty input or if your input contain some illegal characters like `'\0'`, CRLs or `'\n'`/`\r'`, through it will accept any non us-ascii  utf-8 character. (We could also allow the "bad" ascii chars by using encoded words even in a internationalized mail, but they _really_ should not be in there so I still exclude them).

Btw. the `FIXME` refers to the fact that a sequence of more then one whitespace is collapsed into a single whitespace. It never removes them completely. There are a number of reasons for this which don't really apply any more due to how some parts of the library have changed, through some collapsing can happen anyway even if the FIXME happens due to limitations in the line braking algorithm (but you would have to include something like way over 50 consecutive whitespace).

As a sidenote, does your workaround causes `_` to be treated as ` `? If so then this is on the side of AWS as I'm pretty sure I'm not doing anything like that. Also how do you feed the whitespace into the
type? Do you happen to use thinks like `Word::try_from(" wordwithspace ")`? If so please tell me as this could be either indicate a but in my library or some mail software. But if it's a common mail software it's still a good idea to potentially not create headers which hit their bug (or are just not well interpreted by them).

**Comment by kcking** [comment]
Thanks for the super detailed reply!

I am using `Phrase::new` to construct a display name, and `Email::try_from` to construct an email, then creating a `Mailbox` out of those and encoding to a string using `EncodingBuffer`.

The inputs that are giving trouble are things like `TestCo, Inc.`, which is encoded as `=?utf8?Q?TestCo=2C?= =?utf8?Q?Inc=2E?=`. AWS SES renders this display name as `TestCo,Inc.` (w/o a space). Should the header encoding be adding any spaces inside the encoding of the words? From glancing through the spec, it looks like the space between the two encoded words isn't supposed to be rendered in the output, but is just interpreted as the concatenation of the adjacent encoded words.

Adding a pre-processing step that replaces ` ` with `_` ends up encoding the same input as `=?utf8?Q?TestCo=2C_Inc=2E?=`, which AWS renders as `TestCo, Inc.`

Unfortunately, if there isn't a special character like a `,` in the input, then equals encoding isn't used and the underscore is sent to AWS SES and rendered as an underscore instead of a space, so my workaround is insufficient.

I will try splitting into `Word`s manually, but would love any additional advice you have on this!

**Comment by rustonaut** [comment]
Thanks, It's surly a bug/edge case I overlooked. Basically it's the
encoding algorithm overlooking that there happens to be another encoded
word after the space. I will fix that soon, through I will have to look
into the spec. As you said space between encoded words should be ignored,
but if you place two spaces then if I remember correctly both are not
ignored...

Independently of this I should change the algorithm to preferably encoded
the whole phrase as one thing instead of word by word (so the part about a
phrase consisting of a vec of words will likely go away).

**Comment by kcking** [comment]
Sounds good! In the meantime encoding each word individually and prepending a space on all but the first is a workaround I have found. Thanks for the help!
